File: painradar/problem.py
"""Extract a concise problem statement out of a post's title and body."""
from __future__ import annotations

import re

from .signals import detect_pain_signals

_SENTENCE_SPLIT_RE = re.compile(r"(?<=[.!?])\s+")
MAX_LENGTH = 220


def _split_sentences(text: str) -> list[str]:
    text = text.strip()
    if not text:
        return []
    return [s.strip() for s in _SENTENCE_SPLIT_RE.split(text) if s.strip()]


def _truncate(text: str, max_length: int = MAX_LENGTH) -> str:
    if len(text) <= max_length:
        return text
    return text[: max_length - 1].rsplit(" ", 1)[0] + "…"


def extract_problem_statement(title: str, selftext: str) -> str:
    """Pick the most informative sentence describing the pain, defaulting to the title."""
    for sentence in _split_sentences(selftext):
        if detect_pain_signals(sentence).matched_categories:
            return _truncate(sentence)
    return _truncate(title.strip())

File: src/problem.ts
import { detectPainSignals } from "./signals.js";

const SENTENCE_SPLIT_RE = /(?<=[.!?])\s+/;
const MAX_LENGTH = 220;

function splitSentences(text: string): string[] {
  const trimmed = text.trim();
  if (!trimmed) {
    return [];
  }
  return trimmed
    .split(SENTENCE_SPLIT_RE)
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
}

function truncate(text: string, maxLength: number = MAX_LENGTH): string {
  if (text.length <= maxLength) {
    return text;
  }
  const cut = text.slice(0, maxLength - 1);
  const lastSpace = cut.lastIndexOf(" ");
  const base = lastSpace >= 0 ? cut.slice(0, lastSpace) : cut;
  return base + "…";
}

export function extractProblemStatement(title: string, selftext: string): string {
  for (const sentence of splitSentences(selftext)) {
    if (detectPainSignals(sentence).matchedCategories.size > 0) {
      return truncate(sentence);
    }
  }
  return truncate(title.trim());
}
